qutip/qobjevo.py

```diff
@@ -0,0 +1,1354 @@
+                op[0] = other * op[0]
+            return res
+        else:
+            res *= other
+            return res
+
+    def __imul__(self, other):
+        if isinstance(other, Qobj) or isinstance(other, Number):
+            self.cte *= other
+            for op in self.ops:
+                op[0] *= other
+        elif isinstance(other, QobjEvo):
+            if other.const:
+                self.cte *= other.cte
+                for op in self.ops:
+                    op[0] *= other.cte
+            elif self.const:
+                cte = self.cte.copy()
+                self = other.copy()
+                self.cte = cte * self.cte
+                for op in self.ops:
+                    op[0] = cte*op[0]
+            else:
+                cte = self.cte.copy()
+                self.cte *= other.cte
+                new_terms = []
+                old_ops = self.ops
+                if not other.dummy_cte:
+                    for op in old_ops:
+                        new_terms.append(self._ops_mul_cte(op, other.cte, "R"))
+                if not self.dummy_cte:
+                    for op in other.ops:
+                        new_terms.append(self._ops_mul_cte(op, cte, "L"))
+
+                for op_left in old_ops:
+                    for op_right in other.ops:
+                        new_terms.append(self._ops_mul_(op_left,
+                                                        op_right))
+                self.ops = new_terms
+                self.args.update(other.args)
+                self.dummy_cte = self.dummy_cte and other.dummy_cte
+                self.N_obj = (len(self.ops) if
+                              self.dummy_cte else len(self.ops) + 1)
+            self._reset_type()
+
+        else:
+            raise TypeError("td_qobj can only be multiplied" +
+                            " with td_qobj, Qobj or numbers")
+        return self
+
+    def __div__(self, other):
+        if isinstance(other, (int, float, complex,
+                              np.integer, np.floating, np.complexfloating)):
+            res = self.copy()
+            res *= other**(-1)
+            return res
+        else:
+            raise TypeError('Incompatible object for division')
+
+    def __idiv__(self, other):
+        if isinstance(other, (int, float, complex,
+                              np.integer, np.floating, np.complexfloating)):
+            self *= other**(-1)
+        else:
+            raise TypeError('Incompatible object for division')
+        return self
+
+    def __truediv__(self, other):
+        return self.__div__(other)
+
+    def __neg__(self):
+        res = self.copy()
+        res.cte = (-res.cte)
+        for op in res.ops:
+            op[0] = -op[0]
+        return res
+
+    def _ops_mul_(self, opL, opR):
+        new_f = _Prod(opL[1], opR[1])
+        new_op = [opL[0]*opR[0], new_f, None, 0]
+        if opL[3] == opR[3] and opL[3] == 2:
+            new_op[2] = "("+opL[2]+") * ("+opR[2]+")"
```

Comment:
op[2] are they original function, so string, callable or array, op[3] are the the type identifier. op[0] is the Qobj and op[1] is a callable version of the function. self.ops should be a list of tuple (or even it's own class) not a list of list now that I think about it.

Reply:
I rewrote it with namedtuple instead of list to make it more readable.
I am also using enum or string for flags instead of number. I will push those change when I am done with the multi compiled flags.